### ISSUE TITLE
Nested transactions compete with on a single connection

### ISSUE THREAD

**Issue by d11wtq** [Vincit/objection.js#838]
This may be a Knex issue and not an Objection issue. I may also be using the automatic SAVEPOINT feature incorrectly. This arises when using Promise.all to resolve a series of asynchronously executed queries that each run in transactions, if the query context is already inside a top-level transaction.

``` javascript
const Objection = require('objection');
const Model = require('./someModel');

async function causeBug() {
  await Objection.transaction(
    Model.knex(),
    async trx => { // BEGIN
      await Promise.all(
        [1, 2, 3].map(
          async num => {
            await Objection.transaction(
              trx,
              async savepointTrx => { // SAVEPOINT x
                await Model.query(savepointTrx).insert({num: num});
              } // RELEASE SAVEPOINT x
            );
          }
        )
      );
    } // COMMIT
  );
}
```

This results in "No such SAVEPOINT" most of the time for me, as each of the promises started in the loop uses the same connection and they all do SAVEPOINT sp1, SAVEPOINT sp2 etc in order, then each of their insert queries follows, then finally they try to release the SAVEPOINTs out of order.

I can work around it by changing Promise.all to a for..of loop with await inside each iteration of the loop so the queries happen sequentially, but I'd have expected the client to handle that sort of logic for me by using a resource pool.

Is this expected behavior?

**Comment by d11wtq** [comment]
Typing this on a phone so not easy to add detail, but if more detail is required I can post better examples and a test case in a few hours.

**Comment by d11wtq** [comment]
Note that I'm using Node's built-in Promise. No transpiler.

**Comment by d11wtq** [comment]
Looking at the code, it appears this would be a knex issue. The transaction is just a knex transaction.

**Comment by koskimas** [comment]
They have to use the same connection. Transactions are always connection specific. This is not even a knex issue.

**Comment by d11wtq** [comment]
They do, but they shouldn't be allowed to query the connection simultaneously. If I start 10 queries in parallel they should each have to wait for the connection to be free (i.e. not in use by another query). It's certainly fixable. Basically the Knex code inplement acquireConnection to always resolve itself unconditionally for a transaction, whereas it should have similar semantics to trying to access a single connection from a connection pool of size 1. In that case the above code would not blow up with an error, as each query would only be able to run if none of the others are running at the same time.

**Comment by elhigu** [comment]
Problem above is not about using multiple connections, but you are startin mutiple nested transactions parallel (making then deeply nested) and then you are resolving them in wrong order... knex probably could prevent wait for all inner transactions to resolve before outer is resolved/committed, but in that case also your code would be working different way that you think it is working.

**Comment by elhigu** [comment]
Sorry, I was on mobile during the earlier answer so I didn't have patience to write properly. I also thought a bit more about this in bus.

In this case queries are running probably more or less like this (since your code is not waiting that earlier nested transaction is completed, before starting new one):

```
-- start savepoint 1
-- start savepoint 2
-- start savepoint 3
-- insert 1
-- insert 2
-- insert 3
-- end savepoint 1
-- end savepoint 2
-- end savepoint 3
```

So if you would be doing more queries in those separate nested transactions, pretty much all of them would be ran in the most nested one, even if knex would resolve `end savepoint X` stuff in correct order.

I believe this actually could be fixed in knex, but it would require that each nested transaction would spawn new separate `nestedtrx` object and keep track that if there is nested transaction started from parent, then all the queries put to that parent transaction must be queued until the nested transaction has completed. 

That way if one tries to start multiple nested transactions from the same parent, each one of them would still be ran one at the time.

However I think that this feature is so marginal and easy to handle in client side, that I don't believe any of current knex contributors implementing it any time soon (there are lots of more critical issues to be fixed). Would be fine feature request though if some day we are running out of issues in knex.

**Comment by d11wtq** [comment]
Thanks for coming back to this with more detail.

I was going to have a go at solving this locally and if I got something working then I'd put up a PR for Knex. You're right... that's exactly the order the queries happen in and of course, postgresql complains that SAVEPOINT 2 does not exist (because SAVEPOINT 1 was already released and was started before it).

Since I have worked around it too it's also not a priority for me, but something I will likely look at when I have some free time.
